WebAudio: Limit the number of realtime audio threads, and constrain their runtime

jernoble opened the pull request:
#### 02ffbc50d8a6ce4d67e41c7137f58f0ff1f883ec
<pre>
WebAudio: Limit the number of realtime audio threads, and constrain their runtime
<a href="https://bugs.webkit.org/show_bug.cgi?id=263074">https://bugs.webkit.org/show_bug.cgi?id=263074</a>
rdar://116864442

Reviewed by Chris Dumez.

In 268047@main, RemoteAudioDestinationProxy threads were made to run in realtime mode, thereby
avoiding audible stutters and dropouts in periods of high thread contention. However realtime
threads come at a cost, and creating too many realtime threads is problematic from a kernel
scheduling perspective.

Limit the number of concurrent realtime audio threads that can be created to 3. This should allow
well behaved sites to use a small number of concurrent AudioContexts (e.g., with different
latencies and sample rates), while still mitigating the impact of degenerate sites making tens or
even hundreds of AudioContexts.

Additionally, adopt threading APIs (on Darwin OSs, for now) to constrain the runtime characteristics
of these threads. These time constraints will inform the kernel scheduler what the predicted
periodicty and runtime duration of these realtime threads are, and these limits are derivable from
the size of the WebAudio rendering quantum size and its sample rate. When these limits are
exceeded, the scheduler will de-prioritize the realtime thread, reverting it to a standard QOS
thread.

Together, these two limits will mitigate the impact of creating many realtime threads, as well as
mitigating the impact of a small number of poorly behaving realtime threads.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/Threading.h:
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::Thread::setThreadTimeConstraints):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::startRenderingThread):
(WebKit::RemoteAudioDestinationProxy::stopRenderingThread):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:

Canonical link: <a href="https://commits.webkit.org/269291@main">https://commits.webkit.org/269291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37f35b00745a7bdad3ca4bb964c020d6952ac9a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22289 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20439 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26539 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22608 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21493 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19159 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24815 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19079 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20013 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26263 "layout-tests (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19194 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20099 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24130 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21451 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20666 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17598 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25495 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20020 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6010 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5268 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24227 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26770 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20616 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5842 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
```diff
@@ -92,7 +94,27 @@ void RemoteAudioDestinationProxy::startRenderingThread()
+    // rather than have separate realtime threads for each RemoteAudioDestinationProxy.
+    bool createRealtimeThread = s_realtimeThreadCount < s_maximumConcurrentRealtimeThreads;
+    if (createRealtimeThread) {
+        m_threadIsRealtime = true;
```
I'd have called this "m_isRealtimeThread" to match our coding style of having a prefix.

cdumez commented on Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
```diff
@@ -92,7 +94,27 @@ void RemoteAudioDestinationProxy::startRenderingThread()
+
+    // FIXME(263073): Coalesce compatible realtime threads together to render sequentially
+    // rather than have separate realtime threads for each RemoteAudioDestinationProxy.
+    bool createRealtimeThread = s_realtimeThreadCount < s_maximumConcurrentRealtimeThreads;
```
Needs a prefix like "shouldCreateRealtimeThread"

jernoble committed:
WebAudio: Limit the number of realtime audio threads, and constrain their runtime
https://bugs.webkit.org/show_bug.cgi?id=263074
rdar://116864442

Reviewed by Chris Dumez.

In 268047@main, RemoteAudioDestinationProxy threads were made to run in realtime mode, thereby
avoiding audible stutters and dropouts in periods of high thread contention. However realtime
threads come at a cost, and creating too many realtime threads is problematic from a kernel
scheduling perspective.

Limit the number of concurrent realtime audio threads that can be created to 3. This should allow
well behaved sites to use a small number of concurrent AudioContexts (e.g., with different
latencies and sample rates), while still mitigating the impact of degenerate sites making tens or
even hundreds of AudioContexts.

Additionally, adopt threading APIs (on Darwin OSs, for now) to constrain the runtime characteristics
of these threads. These time constraints will inform the kernel scheduler what the predicted
periodicty and runtime duration of these realtime threads are, and these limits are derivable from
the size of the WebAudio rendering quantum size and its sample rate. When these limits are
exceeded, the scheduler will de-prioritize the realtime thread, reverting it to a standard QOS
thread.

Together, these two limits will mitigate the impact of creating many realtime threads, as well as
mitigating the impact of a small number of poorly behaving realtime threads.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/Threading.h:
* Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
(WTF::Thread::setThreadTimeConstraints):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::startRenderingThread):
(WebKit::RemoteAudioDestinationProxy::stopRenderingThread):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:

Canonical link: https://commits.webkit.org/269291@main
